File: 507.Perfect Number.c
#include <stdbool.h>

// We define the Perfect Number is a positive integer that is equal to the 
// sum of all its positive divisors except itself.

// Now, given an integer n, write a function that returns true when it is
//  a perfect number and false when it is not.
// Example:
// Input: 28
// Output: True
// Explanation: 28 = 1 + 2 + 4 + 7 + 14
// Note: The input number n will not exceed 100,000,000. (1e8)

bool checkPerfectNumber(int num) {
    if (num <= 1)
        return false;
    
    int high = num / 2;
    int low = 2;
    int acc = 1;
    while(low < high) {
        if(num % low == 0) {
            acc += low;
            high = num / low;
            acc += high;
        }
        ++low;
    }
    return acc == num;
}

File: leetcode/301-600/501-550/507.Perfect Number.go
package main

import "fmt"


func checkPerfectNumber(num int) bool {
    sum := 0
    div := 1
    for div <= num / 2 {
        if num % div == 0 {
            sum += div
        }
        div += 1
    }

    return sum == num
}

func main() {
		fmt.Print(checkPerfectNumber(28))
}
